### DISCUSSION TITLE
SSR: How to fetch some initial data only on inital server side render?

### DISCUSSION THREAD

**Discussion by amuttsch** [vercel/next.js#21057] [Help]
Hi :)

we have a website that makes heavy use of server side rendering. In order to display any page we have to retrieve some site context that determines the content, think of a multi tenant setup. We currently have this context call in the `_app.js` `getInitialProps`. The result is stored in a `useState` so it is persisted on client side page transition. Fetching this data on the client is no option as we do not want a loading spinner to show as the first thing the user sees when they access the page.

`getInitialProps`, however, is executed and returned on *every* page transition, even if it is client side. This causes our application to needlessly fetch and transfer ~100kb of JSON data.

We have seen that after the initial render next.js requests data from pages that implement `getServerSideProps` using the `/_next/data/` routes and return only the data as json. The rendering happens on the client.

Our idea was to check in the `_app.js` `getInintialProps` if we are such a data route and do not execute the fetch call. We have set up a example repository to showcase the issue here: https://github.com/amuttsch/initial-ssr

This is our current check for the data route: 

```
MyApp.getInitialProps =  async (appContext) => {
  const appProps = await App.getInitialProps(appContext);

  // Is website already initialized?
  // FIXME: Is there a better way to distinguish the requests? Or any other way to get data only when rendered on the server?
  if (appContext.ctx.req?.url?.startsWith('/_next/data/')) {
    return {
      ...appProps
    };
  }

  // Load some initial context
  const siteContext = await fetchContext();

  return {
    ...appProps,
    pageProps: {
      siteContext: siteContext
    }
  };
}
```

https://github.com/amuttsch/initial-ssr/blob/869bc58cb7dc20f41f01caa9c3c38a5b68728d36/pages/_app.js#L16

Afterwards, we store the result in `useState`: 

```
export const SiteContext = createContext({});

function MyApp({ Component, pageProps }) {
  const [siteContext] = useState(pageProps.siteContext);
  return <SiteContext.Provider value={siteContext}><Component {...pageProps} /></SiteContext.Provider>
}
```

https://github.com/amuttsch/initial-ssr/blob/869bc58cb7dc20f41f01caa9c3c38a5b68728d36/pages/_app.js#L8

And we can access it in our pages, see https://github.com/amuttsch/initial-ssr/blob/869bc58cb7dc20f41f01caa9c3c38a5b68728d36/pages/index.js#L19 and https://github.com/amuttsch/initial-ssr/blob/869bc58cb7dc20f41f01caa9c3c38a5b68728d36/pages/blog.js#L19

with varying `getServerSideProps` for each route.

This solution seems to be a bit hacky, is there a better / preferred way to fetch data only once on the server?

Thank you! :-)

PS: We actually use `redux` to save the site context state, but we opted for a simpler setup for this discussion. If you know how to accomplish this with https://github.com/kirill-konshin/next-redux-wrapper it would help us too!

**Comment by lalit2005** [comment]
> This solution seems to be a bit hacky, is there a better / preferred way to fetch data only once on the server?

`getStaticProps` could be used to fetch data only once during the build and if it is required to prerender many pages, `getStaticPaths` can be used with `fallback` set to `true`.

Hope this helps 🤞

**Reply by amuttsch** [reply]
`getStaticProps` is not fesable for us as the data changes too often for fetching it at build time. `getStaticPaths` does not work as well as the data is distinguished by subdomain and not by path :(

**Reply by lalit2005** [reply]
> `getStaticProps` is not fesable for us as the  data changes too often for fetching it at build time. `getStaticPaths` does not work as well as the data is distinguished by subdomain and not by path :(

How about Incremental Static Regeneration? 
Site can be regenerated as frequent as required and is paused when no requests come.

**Reply by yevyevyev** [reply]
> `getStaticProps` is not fesable for us as the data changes too often for fetching it at build time. `getStaticPaths` does not work as well as the data is distinguished by subdomain and not by path :(
return empty array in getStaticPath
then it will build the page on runtime and cache it properly
use incremental static regeneration for maintaining the latest data state.

**Comment by devknoll** [comment]
`getInitialProps` in `/_app` (or `/_document` iff none of your pages use static generation) is probably what you want:
* Instead of branching on the URL, always fetch and return the data when on the server (I.e. `ctx.req` is set)
* On the client, cache the data when `/_app` is first rendered. Read & return the cached value from `getInitialProps` on the client.

As long as the data never changes after the first render, this will be sufficient. If the data does change (even rarely), you will need a slightly more complex solution.
